File: src/main/java/org/apache/sysml/hops/spoof2/plan/SNodeNary.java
package org.apache.sysml.hops.spoof2.plan;

import java.util.ArrayList;
import java.util.List;

public class SNodeNary extends SNode
{
	public enum NaryOp {
		//unary operations
		NOT, ABS, SIN, COS, TAN, ASIN, ACOS, ATAN, SIGN, SQRT, LOG, EXP, 
		ROUND, CEIL, FLOOR, 
		SPROP, SIGMOID, SELP, LOG_NZ, 
		
		//binary operations
		PLUS, MINUS, MULT, DIV, MODULUS, INTDIV, 
		LESS, LESSEQUAL, GREATER, GREATEREQUAL, EQUAL, NOTEQUAL, 
		MIN, MAX, AND, OR, POW, //LOG (see unary)
		
		//ternary operations
		PLUS_MULT, MINUS_MULT,
		
		//reorg operations
		TRANSPOSE;
		
		public static boolean contains(String value) {
			for( NaryOp nt : values()  )
				if( nt.name().equals(value) )
					return true;
			return false;
		}
	};
	
	private final NaryOp _type;
	private final JoinCondition[] _joins;
	
	public SNodeNary(SNode input, NaryOp type) {
		//for unary ops, schema propagates
		this(input, type, input.getSchema());
	}
	
	public SNodeNary(SNode input, NaryOp type, List<String> schema) {
		super(input);
		_type = type;
		
		//for unary ops, schema propagates
		_joins = new JoinCondition[0];
		_schema.addAll(schema);
	}
	
	public SNodeNary(ArrayList<SNode> inputs, NaryOp type, JoinCondition... joins) {
		super(inputs);
		_type = type;
		
		//for n-ary ops, the schema is the union which contains 
		//the join indexes just once
		_joins = joins;
		for(SNode input : inputs)
			_schema.addAll(input.getSchema());
		for(JoinCondition cond : joins)
			_schema.remove(cond.getRight());
	}
	
	public NaryOp getOp() {
		return _type;
	}
	
	public JoinCondition[] getJoins() {
		return _joins;
	}
	
	public boolean containsJoinCondition(String attr) {
		for( JoinCondition join : _joins )
			if( join.getLeft().equals(attr) || join.getRight().equals(attr) )
				return true;
		return false;
	}

	@Override
	public boolean isIndexAggregator() {
		return false;
	}

	@Override
	public boolean isIndexPropagator() {
		return true;
	}

	@Override
	public String toString() {
		switch( _type ) {
			case PLUS: return "nary(+)";
			case MINUS: return "nary(-)";
			case MULT: return "nary(*)";
			case DIV: return "nary(/)";
			case MODULUS: return "nary(%%)";
			case INTDIV: return "nary(%/%)";
			case LESS: return "nary(<)";
			case LESSEQUAL: return "nary(<=)";
			case GREATER: return "nary(>)";
			case GREATEREQUAL: return "nary(>=)";
			case EQUAL: return "nary(==)";
			case NOTEQUAL: return "nary(!=)";
			case PLUS_MULT: return "nary(+*)";
			case MINUS_MULT: return "nary(-*)";
			case TRANSPOSE: return "nary(t)";
		
			default:
				return "nary("+_type.name().toLowerCase()+")";
		}
	}

	//equi join condition, use multiple for conjunctions
	public static class JoinCondition {
		private final String _leftIndex;
		private final String _rightIndex;
		
		public JoinCondition(String left, String right) {
			_leftIndex = left;
			_rightIndex = right;
		}
		
		public String getLeft() {
			return _leftIndex;
		}
		
		public String getRight() {
			return _rightIndex;
		}
	}
}

File: src/main/java/org/apache/sysml/hops/spoof2/plan/SNodeNary.kt
package org.apache.sysml.hops.spoof2.plan

import com.google.common.collect.HashMultiset
import com.google.common.collect.Multiset
import org.apache.sysml.hops.spoof2.plan.SNodeNary.NaryOp.*

class SNodeNary(
        val op: NaryOp,
        inputs: List<SNode>
) : SNode(inputs) {
    init {
        refreshSchema()
    }

    enum class NaryOp(val commutative: Boolean = false) {
        //unary operations
        NOT,
        ABS, SIN, COS, TAN, ASIN, ACOS, ATAN, SIGN, SQRT, LOG, EXP,
        ROUND, CEIL, FLOOR,
        SPROP, SIGMOID, SELP, LOG_NZ,

        //binary operations
        PLUS(true),
        MINUS, MULT(true), DIV, MODULUS, INTDIV,
        LESS, LESSEQUAL, GREATER, GREATEREQUAL, EQUAL(true), NOTEQUAL(true),
        MIN(true), MAX(true), AND(true), OR(true), POW, //LOG (see unary)

        //ternary operations
        PLUS_MULT,
        MINUS_MULT;

        companion object {
            operator fun contains(value: String): Boolean {
                return values().any { it.name == value }
            }
        }
    }

    override fun shallowCopyNoParentsYesInputs() = SNodeNary(op, inputs)
    override fun compare(o: SNode): Boolean {
        if( o !is SNodeNary || o.op != this.op)
            return false
        if( o.inputs == this.inputs )
            return true
        return false
    }

    fun compareCommutative(o: SNode): Boolean {
        if( o !is SNodeNary || o.op != op)
            return false
        // warning: if the schema is different, need to adjust
        return op.commutative && o.inputs.toSet() == this.inputs.toSet()
    }

    constructor(op: NaryOp, vararg inputs: SNode) : this(op, inputs.asList())

    override fun toString(): String {
        return when (op) {
            PLUS ->         "nary(+)"
            MINUS ->        "nary(-)"
            MULT ->         "nary(*)"
            DIV ->          "nary(/)"
            MODULUS ->      "nary(%%)"
            INTDIV ->       "nary(%/%)"
            LESS ->         "nary(<)"
            LESSEQUAL ->    "nary(<=)"
            GREATER ->      "nary(>)"
            GREATEREQUAL -> "nary(>=)"
            EQUAL ->        "nary(==)"
            NOTEQUAL ->     "nary(!=)"
            PLUS_MULT ->    "nary(+*)"
            MINUS_MULT ->   "nary(-*)"
            else ->         "nary(${op.name.toLowerCase()})"
        }
    }

    override fun checkArity() {
        val expect = when( op ) {
            PLUS, MIN, MAX, MULT, EQUAL, AND, OR -> -1

            POW, MINUS, DIV, MODULUS, INTDIV, LESS, LESSEQUAL,
            GREATER, GREATEREQUAL, NOTEQUAL -> 2

            PLUS_MULT, MINUS_MULT -> 3

            ABS, SIN, COS, TAN, ASIN, ATAN, SIGN, SQRT, LOG, EXP, ROUND, CEIL, FLOOR,
            SPROP, SIGMOID, SELP, LOG_NZ, NOT, ACOS -> 1
        }
        this.check(expect == -1 || expect == inputs.size) {"bad arity for $op: ${inputs.size}"}
    }

    override fun refreshSchema() {
        // union of input schemas
        schema.clear()
        inputs.forEach { schema.unionWithBound(it.schema) }
    }

    /**
     * Get how many times each label is present in the input.
     */
    fun getJoinLabelCounts(): Multiset<Name> {
        val ms = HashMultiset.create<Name>()
        inputs.forEach {
            ms.addAll(it.schema.names)
        }
        return ms
    }
}